crates/primitives/src/contract.rs

```diff
@@ -0,0 +1,143 @@
+//! Helpers for deriving contract addresses
+
+use crate::{keccak256, Address, U256};
+use reth_rlp::Encodable;
+use reth_rlp_derive::RlpEncodable;
+
+/// The address for an Ethereum contract is deterministically computed from the
+/// address of its creator (sender) and how many transactions the creator has
+/// sent (nonce). The sender and nonce are RLP encoded and then hashed with Keccak-256.
+pub fn get_contract_address(sender: impl Into<Address>, nonce: impl Into<U256>) -> Address {
+    #[derive(RlpEncodable)]
+    struct S {
+        sender: Address,
+        nonce: U256,
+    }
+    let sender = S { sender: sender.into(), nonce: nonce.into() };
+    let mut buf = Vec::new();
+    sender.encode(&mut buf);
+    let hash = keccak256(buf);
+    let addr: [u8; 20] = (&hash[12..]).try_into().expect("correct len");
+    Address::from(addr)
+}
+
+/// Returns the CREATE2 address of a smart contract as specified in
+/// [EIP1014](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1014.md)
+///
+/// keccak256( 0xff ++ senderAddress ++ salt ++ keccak256(init_code))[12..]
+pub fn get_create2_address(
+    from: impl Into<Address>,
+    salt: impl AsRef<[u8]>,
+    init_code: impl AsRef<[u8]>,
+) -> Address {
+    let init_code_hash = keccak256(init_code);
+    get_create2_address_from_hash(from, salt, init_code_hash)
+}
+
+/// Returns the CREATE2 address of a smart contract as specified in
+/// [EIP1014](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1014.md),
+/// taking the pre-computed hash of the init code as input.
+///
+/// keccak256( 0xff ++ senderAddress ++ salt ++ keccak256(init_code))[12..]
+pub fn get_create2_address_from_hash(
+    from: impl Into<Address>,
+    salt: impl AsRef<[u8]>,
+    init_code_hash: impl AsRef<[u8]>,
+) -> Address {
+    let from = from.into();
+    let salt = salt.as_ref();
+    let init_code_hash = init_code_hash.as_ref();
+
+    let mut bytes = Vec::with_capacity(1 + 20 + salt.len() + init_code_hash.len());
```

Comment:
shouldn't we enforce `salt` to be 32 bytes?

Reply:
good point, will do.
salt is always 32 bytes:
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1014.md